## Libraries

In [2]:
import pandas as pd
import numpy as np

## Analyse exploratoire

In [3]:
df = pd.read_csv("vehicles.csv")

In [4]:
# Afficher les cinq premières lignes du df
df.head()

id                                                url  \
0  7222695916  https://prescott.craigslist.org/cto/d/prescott...   
1  7218891961  https://fayar.craigslist.org/ctd/d/bentonville...   
2  7221797935  https://keys.craigslist.org/cto/d/summerland-k...   
3  7222270760  https://worcester.craigslist.org/cto/d/west-br...   
4  7210384030  https://greensboro.craigslist.org/cto/d/trinit...   

                   region                         region_url  price  year  \
0                prescott    https://prescott.craigslist.org   6000   NaN   
1            fayetteville       https://fayar.craigslist.org  11900   NaN   
2            florida keys        https://keys.craigslist.org  21000   NaN   
3  worcester / central MA   https://worcester.craigslist.org   1500   NaN   
4              greensboro  https://greensboro.craigslist.org   4900   NaN   

  manufacturer model condition cylinders  ... size  type paint_color  \
0          NaN   NaN       NaN       NaN  ...  NaN   NaN         NaN   
1          NaN   NaN       NaN       NaN  ...  NaN   NaN         NaN   
2          NaN   NaN       NaN       NaN  ...  NaN   NaN         NaN   
3          NaN   NaN       NaN       NaN  ...  NaN   NaN         NaN   
4          NaN   NaN       NaN       NaN  ...  NaN   NaN         NaN   

  image_url description county state lat long posting_date  
0       NaN         NaN    NaN    az NaN  NaN          NaN  
1       NaN         NaN    NaN    ar NaN  NaN          NaN  
2       NaN         NaN    NaN    fl NaN  NaN          NaN  
3       NaN         NaN    NaN    ma NaN  NaN          NaN  
4       NaN         NaN    NaN    nc NaN  NaN          NaN  

[5 rows x 26 columns]

In [5]:
# Afficher les cinq dernières lignes du df
df.tail()

id                                                url  \
426875  7301591192  https://wyoming.craigslist.org/ctd/d/atlanta-2...   
426876  7301591187  https://wyoming.craigslist.org/ctd/d/atlanta-2...   
426877  7301591147  https://wyoming.craigslist.org/ctd/d/atlanta-2...   
426878  7301591140  https://wyoming.craigslist.org/ctd/d/atlanta-2...   
426879  7301591129  https://wyoming.craigslist.org/ctd/d/atlanta-2...   

         region                      region_url  price    year manufacturer  \
426875  wyoming  https://wyoming.craigslist.org  23590  2019.0       nissan   
426876  wyoming  https://wyoming.craigslist.org  30590  2020.0        volvo   
426877  wyoming  https://wyoming.craigslist.org  34990  2020.0     cadillac   
426878  wyoming  https://wyoming.craigslist.org  28990  2018.0        lexus   
426879  wyoming  https://wyoming.craigslist.org  30590  2019.0          bmw   

                           model condition    cylinders  ... size       type  \
426875         maxima s sedan 4d      good  6 cylinders  ...  NaN      sedan   
426876  s60 t5 momentum sedan 4d      good          NaN  ...  NaN      sedan   
426877          xt4 sport suv 4d      good          NaN  ...  NaN  hatchback   
426878           es 350 sedan 4d      good  6 cylinders  ...  NaN      sedan   
426879  4 series 430i gran coupe      good          NaN  ...  NaN      coupe   

       paint_color                                          image_url  \
426875         NaN  https://images.craigslist.org/00o0o_iiraFnHg8q...   
426876         red  https://images.craigslist.org/00x0x_15sbgnxCIS...   
426877       white  https://images.craigslist.org/00L0L_farM7bxnxR...   
426878      silver  https://images.craigslist.org/00z0z_bKnIVGLkDT...   
426879         NaN  https://images.craigslist.org/00Y0Y_lEUocjyRxa...   

                                              description county state  \
426875  Carvana is the safer way to buy a car During t...    NaN    wy   
426876  Carvana is the safer way to buy a car During t...    NaN    wy   
426877  Carvana is the safer way to buy a car During t...    NaN    wy   
426878  Carvana is the safer way to buy a car During t...    NaN    wy   
426879  Carvana is the safer way to buy a car During t...    NaN    wy   

              lat       long              posting_date  
426875  33.786500 -84.445400  2021-04-04T03:21:31-0600  
426876  33.786500 -84.445400  2021-04-04T03:21:29-0600  
426877  33.779214 -84.411811  2021-04-04T03:21:17-0600  
426878  33.786500 -84.445400  2021-04-04T03:21:11-0600  
426879  33.779214 -84.411811  2021-04-04T03:21:07-0600  

[5 rows x 26 columns]

In [6]:
# Affciher les lignes du df, aléatoirement
df.sample(n=10)

id                                                url  \
409338  7316413388  https://seattle.craigslist.org/est/ctd/d/belle...   
173776  7304708174  https://louisville.craigslist.org/ctd/d/louisv...   
402629  7315642458  https://kpr.craigslist.org/cto/d/kennewick-maz...   
376527  7314713149  https://houston.craigslist.org/ctd/d/houston-2...   
344639  7307200441  https://charleston.craigslist.org/ctd/d/charle...   
423339  7313416180  https://milwaukee.craigslist.org/ctd/d/oconomo...   
352199  7309177111  https://chattanooga.craigslist.org/ctd/d/merce...   
182248  7311443390  https://baltimore.craigslist.org/ctd/d/la-verg...   
422759  7315113666  https://milwaukee.craigslist.org/cto/d/milwauk...   
424548  7309156416  https://milwaukee.craigslist.org/ctd/d/milwauk...   

                          region                          region_url  price  \
409338            seattle-tacoma      https://seattle.craigslist.org  49498   
173776                louisville   https://louisville.craigslist.org   2999   
402629  kennewick-pasco-richland          https://kpr.craigslist.org  16500   
376527                   houston      https://houston.craigslist.org  13983   
344639                charleston   https://charleston.craigslist.org  47590   
423339                 milwaukee    https://milwaukee.craigslist.org  11550   
352199               chattanooga  https://chattanooga.craigslist.org      0   
182248                 baltimore    https://baltimore.craigslist.org      0   
422759                 milwaukee    https://milwaukee.craigslist.org   1700   
424548                 milwaukee    https://milwaukee.craigslist.org  42990   

          year   manufacturer                         model  condition  \
409338  2018.0           audi                            q7        NaN   
173776  2002.0          volvo                           s80        NaN   
402629  2018.0          mazda                             3  excellent   
376527  2013.0       infiniti                     g37 sedan        NaN   
344639  2020.0            gmc    yukon slt sport utility 4d       good   
423339  2013.0      chevrolet               traverse lt awd        NaN   
352199  2018.0  mercedes-benz                       e-class        NaN   
182248  2014.0            NaN         Freightliner Cascadia        NaN   
422759  1998.0        pontiac                       sunfire        NaN   
424548  2015.0            bmw  6 series 650i convertible 2d       good   

          cylinders  ...         size   type paint_color  \
409338  6 cylinders  ...     mid-size    SUV       black   
173776          NaN  ...          NaN  sedan         NaN   
402629  4 cylinders  ...    full-size  sedan      silver   
376527  6 cylinders  ...     mid-size  sedan      custom   
344639  8 cylinders  ...          NaN    SUV       white   
423339  6 cylinders  ...          NaN  wagon        grey   
352199  6 cylinders  ...  sub-compact  coupe       white   
182248          NaN  ...          NaN    NaN         NaN   
422759  4 cylinders  ...          NaN  coupe       green   
424548          NaN  ...          NaN  other       black   

                                                image_url  \
409338  https://images.craigslist.org/00c0c_co8v1ZAjtt...   
173776  https://images.craigslist.org/00J0J_SiHotUpm23...   
402629  https://images.craigslist.org/00303_eiWdhKya7x...   
376527  https://images.craigslist.org/00505_lXoUFdT0Si...   
344639  https://images.craigslist.org/00U0U_7zjXpDb9cS...   
423339  https://images.craigslist.org/00a0a_6lFzdBGJYZ...   
352199  https://images.craigslist.org/00m0m_fnHezA4L0I...   
182248  https://images.craigslist.org/00M0M_aVuQx61rdc...   
422759  https://images.craigslist.org/00F0F_4qtj8R6jvh...   
424548  https://images.craigslist.org/00N0N_1xMPvfxRAI...   

                                              description county state  \
409338  Most common questions about this vehicle:  Wan...    NaN    wa   
173776  2002 Volvo S80 T6 Executive 4dr Turbo 

In [7]:
# Dimension du df c-a-d le nombre de lignes et colonnes
df.shape

(426880, 26)

In [8]:
# Nombre de colonnes dans df
df.columns

Index(['id', 'url', 'region', 'region_url', 'price', 'year', 'manufacturer',
       'model', 'condition', 'cylinders', 'fuel', 'odometer', 'title_status',
       'transmission', 'VIN', 'drive', 'size', 'type', 'paint_color',
       'image_url', 'description', 'county', 'state', 'lat', 'long',
       'posting_date'],
      dtype='object')

In [9]:
# Le type de données de chaque colonne du df
df.dtypes

id                int64
url              object
region           object
region_url       object
price             int64
year            float64
manufacturer     object
model            object
condition        object
cylinders        object
fuel             object
odometer        float64
title_status     object
transmission     object
VIN              object
drive            object
size             object
type             object
paint_color      object
image_url        object
description      object
county          float64
state            object
lat             float64
long            float64
posting_date     object
dtype: object

La commande ci-dessous, nous me permet d'avoir des informations sur le dataset comme le type de données et le nombre total des valeurs non nulles dans chaque colonne.

Cela nous permet didentifier les valeurs manquantes et les types de données incorrectes 

In [10]:

df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 426880 entries, 0 to 426879
Data columns (total 26 columns):
 #   Column        Non-Null Count   Dtype  
---  ------        --------------   -----  
 0   id            426880 non-null  int64  
 1   url           426880 non-null  object 
 2   region        426880 non-null  object 
 3   region_url    426880 non-null  object 
 4   price         426880 non-null  int64  
 5   year          425675 non-null  float64
 6   manufacturer  409234 non-null  object 
 7   model         421603 non-null  object 
 8   condition     252776 non-null  object 
 9   cylinders     249202 non-null  object 
 10  fuel          423867 non-null  object 
 11  odometer      422480 non-null  float64
 12  title_status  418638 non-null  object 
 13  transmission  424324 non-null  object 
 14  VIN           265838 non-null  object 
 15  drive         296313 non-null  object 
 16  size          120519 non-null  object 
 17  type          334022 non-null  object 
 18  pain

In [11]:
#Statistiques descriptives pour les colonnes numériques (moyenne, écart-type, min, max, quartiles)
df.describe()

id         price           year      odometer  county  \
count  4.268800e+05  4.268800e+05  425675.000000  4.224800e+05     0.0   
mean   7.311487e+09  7.519903e+04    2011.235191  9.804333e+04     NaN   
std    4.473170e+06  1.218228e+07       9.452120  2.138815e+05     NaN   
min    7.207408e+09  0.000000e+00    1900.000000  0.000000e+00     NaN   
25%    7.308143e+09  5.900000e+03    2008.000000  3.770400e+04     NaN   
50%    7.312621e+09  1.395000e+04    2013.000000  8.554800e+04     NaN   
75%    7.315254e+09  2.648575e+04    2017.000000  1.335425e+05     NaN   
max    7.317101e+09  3.736929e+09    2022.000000  1.000000e+07     NaN   

                 lat           long  
count  420331.000000  420331.000000  
mean       38.493940     -94.748599  
std         5.841533      18.365462  
min       -84.122245    -159.827728  
25%        34.601900    -111.939847  
50%        39.150100     -88.432600  
75%        42.398900     -80.832039  
max        82.390818     173.885502

In [12]:
# Statistiques descriptives pour des colonnes catégorielles (nombre de valeurs uniques, fréquence des plus courantes).
df.describe(include='object')

url    region  \
count                                              426880    426880   
unique                                             426880       404   
top     https://prescott.craigslist.org/cto/d/prescott...  columbus   
freq                                                    1      3608   

                            region_url manufacturer   model condition  \
count                           426880       409234  421603    252776   
unique                             413           42   29667         6   
top     https://spokane.craigslist.org         ford   f-150      good   
freq                              2988        70985    8009    121456   

          cylinders    fuel title_status transmission                VIN  \
count        249202  423867       418638       424324             265838   
unique            8       5            6            3             118264   
top     6 cylinders     gas        clean    automatic  1FMJU1JT1HEA52352   
freq          94169  356209       405117       336524                261   

         drive       size    type paint_color  \
count   296313     120519  334022      296677   
unique       3          4      13          12   
top        4wd  full-size   sedan       white   
freq    131904      63465   87056       79285   

                                                image_url  \
count                                              426812   
unique                                             241899   
top     https://images.craigslist.org/00N0N_1xMPvfxRAI...   
freq                                                 7357   

                                              description   state  \
count                                              426810  426880   
unique                                             360911      51   
top     35 VEHICLES PRICED UNDER $3000!!!  BIG TIME! T...      ca   
freq                                                  231   50614   

                    posting_date  
count                     426812  
unique                    381536  
top     2021-04-23T22:13:05-0400  
freq                          12

## Nettoyage

In [148]:
df = pd.read_csv("vehicles.csv")

### Suppression des colonnes inutiles

In [149]:
df.drop(columns=['url','region_url','image_url','description'], inplace=True)

In [150]:
df.columns

Index(['id', 'region', 'price', 'year', 'manufacturer', 'model', 'condition',
       'cylinders', 'fuel', 'odometer', 'title_status', 'transmission', 'VIN',
       'drive', 'size', 'type', 'paint_color', 'county', 'state', 'lat',
       'long', 'posting_date'],
      dtype='object')

### Traitement des valeurs incohérentes

In [151]:
#Suppression des prix incohérents
df = df[(df['price'] >= 1000) & (df['price'] <= 250000)]
df.shape

(380470, 22)

In [152]:
#Suppression des années incohérentes
df = df[(df['year'] >= 1950) & (df['year'] <= 2025)]
df.shape

(378185, 22)

In [153]:
#Suppression des kilométrages incohérents
df = df[(df['odometer'] >= 100) & (df['odometer'] <= 500000)]
df.shape

(371031, 22)

### Gestion des valeurs manquantes

In [154]:
#Suppression des lignes avec des valeurs manquantes importantes
colonnes_importantes = ['price', 'year', 'manufacturer', 'model', 'odometer']
df.dropna(subset=colonnes_importantes, inplace=True)
df.shape

(354200, 22)

### Correction des types de données

In [155]:
# Conversion de la colonne 'posting_date' en format datetime sans les heures
df['posting_date'] = pd.to_datetime(df['posting_date'], errors='coerce', utc=True)
df['posting_date'] = df['posting_date'].dt.tz_convert(None).dt.normalize()

In [156]:
df.sample(n=10)

id                    region  price    year manufacturer  \
24404   7310943884                     chico  22995  2017.0         ford   
250897  7311275109                central NJ   7900  2009.0       subaru   
376816  7314469873                   houston  29995  2017.0         jeep   
110534  7306357216        sarasota-bradenton   4500  2008.0       nissan   
411929  7303948243                 wenatchee   4995  2006.0       saturn   
419977  7315285686                   madison  13900  1994.0    chevrolet   
411302  7311450530                 wenatchee  42995  2017.0          ram   
123286  7314933903                   atlanta  24995  2011.0         ford   
405219  7304628623  kennewick-pasco-richland   8998  2012.0      hyundai   
231104  7303568571                 asheville  10990  2013.0       nissan   

                     model  condition    cylinders    fuel  odometer  ...  \
24404     explorer xlt 4wd        NaN          NaN     gas   99653.0  ...   
250897             impreza        NaN  4 cylinders     gas   64760.0  ...   
376816  wrangler unlimited        NaN  6 cylinders     gas   69677.0  ...   
110534              sentra  excellent  4 cylinders     gas  128000.0  ...   
411929                 ion        NaN          NaN     gas   93797.0  ...   
419977            c/k 1500       good  8 cylinders     gas  144848.0  ...   
411302                1500  excellent  8 cylinders     gas   46638.0  ...   
123286                f250       good  8 cylinders  diesel  171399.0  ...   
405219              tucson        NaN          NaN     gas  124675.0  ...   
231104   sentra s sedan 4d       good          NaN     gas   81557.0  ...   

                      VIN drive       size   type paint_color county state  \
24404                 NaN   4wd        NaN    SUV         NaN    NaN    ca   
250897  JF1GH636X9H826983   4wd        NaN  wagon        grey    NaN    nj   
376816                NaN   4wd   mid-size    SUV         red    NaN    tx   
110534  3N1AB61E18L745336   fwd  full-size  sedan      silver    NaN    fl   
411929  1G8AJ55FX6Z157867   NaN    compact  sedan         red    NaN    wa   
419977  1GCDC14KXRZ123310   rwd  full-size  truck         red    NaN    wi   
411302                NaN   4wd        NaN    NaN         NaN    NaN    wa   
123286                NaN   4wd  full-size  truck        blue    NaN    ga   
405219  KM8JU3ACXCU443025   fwd        NaN    NaN         NaN    NaN    wa   
231104  3N1AB7AP2DL658501   fwd        NaN  sedan         red    NaN    nc   

              lat        long  posting_date  
24404   40.175298 -122.233025    2021-04-22  
250897  40.408250  -74.372882    2021-04-23  
376816  29.680250  -95.404913    2021-04-29  
110534  27.093300  -82.449800    2021-04-13  
411929  47.912956 -122.235725    2021-04-09  
419977  42.189800  -91.775800    2021-05-01  
411302  46.999600 -120.516300    2021-04-23  
123286  33.796981  -84.643779    2021-04-30  
405219  46.213805 -119.161041    2021-04-10  
231104  35.570000  -82.540000    2021-04-08  

[10 rows x 22 columns]

### Gestion des doublons

#### Premier dédoublonnage en utilisant l'immatriculation du véhicule 'VIN' et la date du posting 'posting_date'

In [157]:
df.shape

(354200, 22)

In [158]:
# Exclure les lignes sans VIN (on ne veut pas les traiter comme doublons)
df_vin_non_null = df[df['VIN'].notna() & (df['VIN'] != '')]
df_vin_non_null.shape

(222816, 22)

In [159]:
# Dédoublonnage sur (VIN + posting_date) parmi les lignes valides
df_vin_dedoublonne = df_vin_non_null.drop_duplicates(subset=['VIN', 'posting_date'])
df_vin_dedoublonne.shape

(145799, 22)

In [160]:
# Garder les autres lignes non concernées (avec VIN vide ou NaN)
df_sans_vin = df[df['VIN'].isna() | (df['VIN'] == '')]
df_sans_vin.shape

(131384, 22)

In [161]:
# Recombiner les deux
df = pd.concat([df_vin_dedoublonne, df_sans_vin], ignore_index=True)
df.shape

(277183, 22)

#### Deuxième dédoublonnage sur les lignes sans 'VIN' en utilisant les variables importantes et la date du posting 'posting_date'

In [162]:
# Définir les colonnes clés pour le second dédoublonnage
colonnes_importantes = [
    'region', 'price', 'year', 'manufacturer', 'model', 'posting_date'
]
df.shape

(277183, 22)

In [163]:
# Appliquer le drop_duplicates
df = df.drop_duplicates(subset=colonnes_importantes)
df.shape

(275689, 22)

Conclusion du dédoublonnage :
- 77 017 doublons supprimés en utilisant les colonnes 'VIN' + 'Posting_date' (~35%), ce qui montre un très fort taux de repost d'annonce des mêmes voitures dans la même journée
- 1 494 doublons supprimés sur en utilisant les colonnes importantes (~1%)
Cette suppression évite d'avoir des annonces identiques ou trop proches pour augmenter la qualité de nos analyses

## Axes d'analyse & Visualisation

## Vectorisation